Migrate route and travel time logic

rachelekm opened the pull request:
## Overview

Migrates state and selector logic from the old codebase, refactoring with the latest Taui source code as needed, to support generating neighborhood trip routes and commute times for each network mode. Largely follows state management plan proposed in #685 up to generating rankings.

### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

## Testing Instructions

 * First go back in commit history to get old values to compare against
     *  `git checkout 6f17e33e875b9f8fc215936d45387805cb74bafb` (Commit before removing old UI)
     *  `scripts/update`
     * Update local .env so that the network url var does not have the `VITE_` preface
     * `scripts/server`
 * Login and add the destination address 700 Boylston St. and set the network mode to "Peak"
 * Open up the console to view the networks state and note shape and values of `state.data.networks[index of "peak"].paths`, `state.data.networks[index of "peak"].pathsPerTarget`, `state.data.networks[index of "peak"].targets`,  and `state.data.networks[index of "peak"].travelTimeSurface` (These values are the parsed results of the path/time fetches).
 * Add the following line[ somewhere in the render of `Application` component](https://github.com/azavea/echo-locator/blob/6f17e33e875b9f8fc215936d45387805cb74bafb/src/frontend/src/components/application.js#L166-L199), then take note of shape and values of derived travel times in console output:
     * `console.log(p.neighborhoodTravelTimes)`
 * Return back to this branch:
     *  Adjust .env back
     * `scripts/update`
     * `scripts/server`
 * Open the console and view the networks state, compare the `paths`, `targets`, `pathsPerTarget`, and `targets` shape and values from the ones noted from the old codebase. Confirm they are the same.
 * Add the below snippet to the `Discover` page component to view selector outputs:
 ```
const times = useSelector(selectNeighborhoodTravelTimes);
    const routes = useSelector(selectNeighborhoodRouteGeoJson);
    if (networks && networks[activeMode].timesAndPathsDataReady) {
        console.log("routes", routes);
        console.log("times", times);
    }
 ```
 * Confirm the times output matches the shape and values from the old codebase's `neighborhoodTravelTimes` output.
 * Copy the feature collection from the routes output and input it in a visualizer like https://geojson.io/
 * Open the production echo app and input "700 Boylston St" as the address and set the mode to "Peak"
 * Confirm the rendered routes between the two look the same

Resolves #655

rachelekm committed:
Add initial networks reducer

rachelekm committed:
Migrate times/paths source code

rachelekm committed:
Add neighborhoodTravelTimes selector

rachelekm committed:
Add networkNeighborhoodRoute selector

rachelekm committed:
Add drawNeighborhoodRoute selector

rachelekm committed:
Refactor to latest Taui for route paths

rachelekm committed:
Add new selectors to reducer

rachelekm committed:
Update CHANGELOG

rachelekm commented on src/frontend/src/reducers/neighborhoods/neighborhoodsSlice.ts:
```diff
@@ -18,18 +16,6 @@ export const neighborhoodSlice = createSlice({
-            { payload: neighborhoodBounds }: { payload: NeighborhoodBounds }
-        ) => {
-            state.neighborhoodBounds = neighborhoodBounds;
-        },
```
Took these out because I realized they're set via the `exraReducers` below and there's no scenario where we'd be updating neighborhoods/bounds data from the frontend.

aaronxsu replied:
Agreed. We should only fetch them once when the app initializes.

rachelekm committed:
Standardize inline comment cites

rachelekm commented on src/frontend/src/reducers/neighborhoods/selectors/allNeighborhoodTravelTimes.ts:
```diff
@@ -0,0 +1,51 @@
+// Refactored from old codebase neighborhoodTravelTimes at commit 6f17e33
+
```
This is refactored from the old selector to support travel times for all networks modes up front so we can create the commute time ranges for each neighborhood

rachelekm commented on src/frontend/src/reducers/neighborhoods/selectors/drawNeighborhoodRoute.ts:
```diff
@@ -0,0 +1,140 @@
+// Refactored from old codebase at latest commit before introducing listings: 8ca3b94
+// New code copied from latest Taui v3 to support routes with MapLibre
+
```
This selector has changed significantly from the old one and is now largely the updated logic from Taui, which shifted from leaflet to mapbox. New Taui logic generates geojson's that better fit our maplibre map. When we're ready to add routes we should follow a [similar styling and application from Taui.](https://github.com/conveyal/taui/blob/dev/src/map/geojson-routes.js)

rachelekm commented on src/frontend/src/reducers/networks/utils/parsePathsData.ts:
```diff
@@ -0,0 +1,101 @@
+// Refactored from latest Taui v3 parsePathsData to support new paths data format
+
```
This updated code [ is from this comment](https://github.com/azavea/echo-locator/issues/628#issuecomment-3184615811), written when migrating to new analysis files in #628

aaronxsu replied:
Looks like some trial and error efforts were invested 👏🏼 Good job to make this work for the new data!

rachelekm commented on src/frontend/src/store/store.ts:
```diff
@@ -6,7 +6,10 @@ import rootReducer from "./rootReducer";
+    middleware: getDefaultMiddleware =>
+        getDefaultMiddleware({
+            serializableCheck: false,
+        }).concat(logger),
```
Added to support the paths and times data as array buffers in state

rachelekm commented:
This PR is following the same implementation as our old codebase and Taui: 1. get network analysis details for each mode, 2. on origin change, fetch and parse new paths and times data for each mode 3. Derive all needed route and travel time information in selectors once each state network object in updated with new time/path data.

In bringing things over I realized some cons to keeping this approach:
 - The parsed times and paths data are large binary array buffers that make development difficult. They are not serializable, cause issues in logging and break redux dev tools.
 - Since we're deriving all of our ranking / routing data from selectors we don't have access to ranked neighborhood lists, neighborhood travel times, or neighborhood route in state. This makes debugging more difficult and if we're calling selectors from different components we may be re-running calculations as[ createSelector only memoizes the most recent set of parameters.](https://redux.js.org/usage/deriving-data-selectors#createselector-behavior)
 - Whenever a user switches between addresses the paths and times data will be fetched and everything re-calculated, even if previously fetching/calculated for that address.

While migrating things over I mulled over keeping the networks `request.json` and `transitive.json` in state and then separating out the times and paths, deriving necessary route and travel time information asynchronously within the thunk vs as selectors. This would allow us to throw away those large array buffers and set traveltime and route directly to state. We could also potentially wrap the thunk dispatch in a useMemo. This approach might need some more thought, but I think still might be a worthwhile change. Curious on your thoughts, regardless I think that would be a followup issue so that we can focus on the bringing over the needed logic in this card.

aaronxsu commented on src/frontend/src/reducers/networks/types.ts:
```diff
@@ -0,0 +1,182 @@
+export type NetworkModeOptions =
```
I could see them being enums. I used `enum` directly in another recent project but typescript/linter complained, and thus I switched to: https://dev.to/ivanzm123/dont-use-enums-in-typescript-they-are-very-dangerous-57bh. Up to you.

rachelekm replied:
Cool this was a good suggestion! Added bonus is that this helps bring the dashed syntax from the file paths and the formatted labels to be displayed together using one consistent key syntax. That work to switch this over to enums is in 6866f4c286852f0d34b0f23050c1315feac04fdd

aaronxsu commented on src/frontend/src/reducers/networks/types.ts:
```diff
@@ -0,0 +1,182 @@
+/**
+ * Route data types from old codebase
+ */
+export type LonLat = { lat: number; lon: number };
```
The type looks the same as `originPoint` above. Is it possible to use one of them and remove the other?

rachelekm replied:
Good catch, I created `originPoint` and then `LonLat` was copied over from the source code. I didn't notice these were the same, I just updated so that `LonLat` is the type used throughout the frontend and scrapped `originPoint` in 87a132286b8e28e4aa758004356e38329f5c6d72

aaronxsu commented on src/frontend/src/reducers/networks/types.ts:
```diff
@@ -0,0 +1,182 @@
+    stops: TransitiveStop[];
+};
+
+export type QualifiedLeg = [TransitiveStop, TransitivePattern, TransitiveStop]; // [boardStopId, Pattern, alightStopId]
```
I am confused by the code comment here- are they meant to be here?

rachelekm replied:
These comments were copied from the source code purely as added context but I noticed were not actually even using `QualifiedLeg` or `QualifiedPath` anymore so I took just removed them in 87a132286b8e28e4aa758004356e38329f5c6d72

aaronxsu commented on src/frontend/src/reducers/networks/types.ts:
```diff
@@ -0,0 +1,182 @@
+
+export type TransitivePattern = {
+    pattern_id: string;
+    patterns?: TransitivePattern[];
```
Ah, interesting that we have a recursive type!

aaronxsu commented on src/frontend/src/reducers/networks/networksSlice.ts:
```diff
@@ -0,0 +1,85 @@
+export const selectAllNetworksDataReady = createSelector(
+    [(state: RootState) => state.networks.networks],
+    networks =>
+        networks &&
```
I'd suggest we have `!!netwoks` here instead, so that the type of the return here is purely `boolean` rather than `null | boolean`.

aaronxsu commented on src/frontend/src/reducers/networks/networksThunk.ts:
```diff
@@ -0,0 +1,107 @@
+// Refactored from old codebase loadDataset and fetchAllTimesAndPathsForCoordinate at commit 6f17e33
+// New source code copied from latest Taui v3
```
Good idea to call it out here. Since Taui is under MIT license, it should be permissive. I'd suggest researching if there is anything additional we should include as well.

rachelekm replied:
There doesn't seem to be official guidance other than ensuring that the license is brought forward into the new repo. General internet consensus is that a linking citation in the source files is good practice as well. I added more formal citations in these comments and then brought the Taui license over into the frontend dir in f8ea6679cad2aa604a7b6184211edaff36345a41

aaronxsu commented on src/frontend/src/reducers/networks/networksThunk.ts:
```diff
@@ -0,0 +1,107 @@
+                })
+            );
+
+            return networkDataByMode as Networks;
```
Ah I see why this `as` is necessary here, because of the `Promise<any>` in the API method. This is fine for now. But I'd suggest we add a TODO to come back to it later when budget allows.

rachelekm replied:
Added a TODO as part of 31dc63304019fb034e154381993743620cc99011

aaronxsu commented on src/frontend/src/reducers/networks/utils/coordinateToIndex.ts:
```diff
@@ -0,0 +1,33 @@
+    };
+}
+
+export function coordinateToIndex(network: any, coordinate: originPoint) {
```
There is a `Network` type, I wonder why `any` is used here instead.

rachelekm replied:
Whoops a leftover when copying this fn from Taui, this is updated the `Network` in 31dc63304019fb034e154381993743620cc99011

aaronxsu commented on src/frontend/src/reducers/networks/utils/parsePathsData.ts:
```diff
@@ -0,0 +1,101 @@
+        // Access and egress modes, each is a single byte
+        // const accessMode = dataView.getInt8(byteOffset);
+        byteOffset += 1;
+        // const egressMode = dataView.getInt8(byteOffset);
```
If `accessMode` and `egressMode` are no longer useful, let's remove them

rachelekm replied:
Sounds good, I simplified this comment to just the needed context in 31dc63304019fb034e154381993743620cc99011

aaronxsu commented on src/frontend/src/reducers/networks/utils/createTransitiveRoutes.ts:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    // Find stop
+    const findStop = (stopId: string) => find(td.stops, ["stop_id", stopId])!;
```
I see the uses of non-null assertion operator `!` a few times in this file. As long as we are certain that the element it's finding (a stop here, and a pattern/route in the below method) exists at runtime, it should be fine. Otherwise, we might see crashes.

rachelekm replied:
I went back in a set up some guardrails so that I could take out all these non-null operators in 31dc63304019fb034e154381993743620cc99011

aaronxsu reviewed (commented):
What a great job migrating all these logic over and the refactoring 👏🏼 

I tested based on testing instructions and it works as advertised. I left some questions/suggestion in code and would like to hear what you think!

>This would allow us to throw away those large array buffers and set traveltime and route directly to state.

I'd prefer this since it would be cleaner. I'd consider it if budget allows or if we need it to have better app performance.

>regardless I think that would be a followup issue so that we can focus on the bringing over the needed logic in this card.

Makes sense to me.

rachelekm committed:
Solidify type declarations

rachelekm committed:
Refactor to network mode enum

rachelekm commented:
Thank you for the thorough review, @aaronxsu! I addressed each one in the individual comments, this is all set for another look! Noting that I rebased on develop so you will need tot force pull this branch and then run `scripts/update` once again.

aaronxsu reviewed (approved):
Thanks for addressing the comments. This is now looking very good 👏🏼 Good to go after a changelog entry 🚀

rachelekm commented:
Thank you!! CHANGELOG entry up in d37fbeb731b3d394b2c77365526454e1837c42d3, will merge shortly
